### PR TITLE
[Yul] Move for loop condition out of body.

### LINKED ISSUE #7424: [yul] optimizer can now increase execution step count of a program by 2x
## Description

We had decided on an upper bound of `N * 1.5` number of execution steps for optimized code (see https://github.com/ethereum/solidity/issues/6476#issuecomment-480237241) that is now exceeded most likely because of #7407 

> Does it make sense to set the bound to some value N when executing the non-optimized version and setting it to N * 1.5 when executing the optimized version? This should make the assumption you name true again in most cases, and the cases where it fails should be inspected manually anyway, because it sounds like the optimizer did a bad job for those.

Opened this issue to discuss this change. My intuition is that break for-loop-condition into `if (iszero(condition)) break` adds an additional step for each loop iteration, so in total optimized code may lead to twice the number of steps.

Example unoptimized code

```
{                   
    for { let i_0 } lt(i_0, 0x60) { i_0 := add(i_0, 0x20) }
    {              
        for { let i_1 } lt(i_1, 0x60) { i_1 := add(i_1, 0x20) }
        {
            for { let i_2 } lt(i_2, 0x60) { i_2 := add(i_2, 0x2

            {
                for { } msize() { }
                { }
            }
        }
    }
}
```

Example optimized code (2x "slower")

```
{
    let i := 0

    for { } 1 { i := add(i, 0x20) }
    {                
        if iszero(lt(i, 0x60)) { break }
        let i_1 := 0 
        for { } 1 { i_1 := add(i_1, 0x20) }
        {            
            if iszero(lt(i_1, 0x60)) { break }
            let i_2 := 0
            for { } 1 { i_2 := add(i_2, 0x20) }
            {
                if iszero(lt(i_2, 0x60)) { break }
                for { } 1 { }
                { if iszero(msize()) { break } }
            }
        }
    }
}
```

One solution is to change these two

https://github.com/ethereum/solidity/blob/34ba617ea6eec0bd4e2e9fd94dcbbb964626955e/test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp#L121

https://github.com/ethereum/solidity/blob/34ba617ea6eec0bd4e2e9fd94dcbbb964626955e/test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp#L104

to

```
(yul::test::yul_fuzzer::yulFuzzerUtil::maxSteps * 2)
```

### PR DESCRIPTION
**Pull request by chriseth**
Closes https://github.com/ethereum/solidity/issues/7424

### CODE REVIEW

File: libyul/optimiser/ForLoopConditionIntoBody.cpp [hunk 1]
```diff
@@ -54,3 +56,45 @@ void ForLoopConditionIntoBody::operator()(ForLoop& _forLoop)
...
+	YulString iszero = m_dialect.booleanNegationFunction()->name;
+	langutil::SourceLocation location = locationOf(*firstStatement.condition);
+
+	if (
+		firstStatement.condition->type() == typeid(FunctionCall) &&
+		boost::get<FunctionCall>(*firstStatement.condition).functionName.name == iszero
+	)
+		_forLoop.condition = make_unique<Expression>(std::move(boost::get<FunctionCall>(*firstStatement.condition).arguments.front()));
+	else
+		_forLoop.condition = make_unique<Expression>(FunctionCall{
```
Comment (ekpyron) [inline libyul/optimiser/ForLoopConditionIntoBody.cpp]: We could always do this - ``iszero(iszero(x))`` -> ``x`` should be simplified for ``if`` conditions elsewhere, right?

File: libyul/optimiser/ForLoopConditionIntoBody.cpp [hunk 2]
```diff
@@ -54,3 +56,45 @@ void ForLoopConditionIntoBody::operator()(ForLoop& _forLoop)
...
+		_forLoop.condition->type() != typeid(Literal) ||
+		valueOfLiteral(boost::get<Literal>(*_forLoop.condition)) == u256(0) ||
+		_forLoop.body.statements.empty() ||
+		_forLoop.body.statements.front().type() != typeid(If)
+	)
+		return;
+
+	If& firstStatement = boost::get<If>(_forLoop.body.statements.front());
+	if (
+		firstStatement.body.statements.size() != 1 ||
```
Comment (ekpyron) [inline libyul/optimiser/ForLoopConditionIntoBody.cpp]: That could actually be ``firstStatement.body.statements.empty()`` but doesn't really matter.

File: libyul/optimiser/ForLoopConditionIntoBody.h [hunk 3]
```diff
@@ -48,4 +48,35 @@ class ForLoopConditionIntoBody: public ASTModifier
...
+ *
+ * for { ... } c { ... } {
+ *   ...
+ * }
+ *
+ * if ``c`` is movable.
+ *
+ * The LiteralRematerialiser should be run before this step.
+ */
+class ForLoopConditionOutOfBody: public ASTModifier
```
Comment (ekpyron) [inline libyul/optimiser/ForLoopConditionIntoBody.h]: Ok, it's small and closely related to the other one, but might it still deserve it's own file :-)?

File: libyul/optimiser/ForLoopConditionIntoBody.h [hunk 4]
```diff
@@ -48,4 +48,35 @@ class ForLoopConditionIntoBody: public ASTModifier
 	Dialect const& m_dialect;
 };
 
+/**
+ * Reverses the transformation of ForLoopConditionIntoBody.
+ *
+ * for { ... } 1 { ... } {
+ *   if iszero(c) { break }
```
Comment (ekpyron) [inline libyul/optimiser/ForLoopConditionIntoBody.h]: That's only half the truth.

File: test/libsolidity/gasTests/abiv2_optimised.sol [hunk 5]
```diff
@@ -17,9 +17,9 @@ contract C {
 // optimize-yul: true
 // ----
 // creation:
-//   codeDepositCost: 624200
-//   executionCost: 657
-//   totalCost: 624857
+//   codeDepositCost: 608400
```
Comment (ekpyron) [inline test/libsolidity/gasTests/abiv2_optimised.sol:20]: Nice!

File: libyul/optimiser/ForLoopConditionIntoBody.cpp (hunk 1 above)
Comment (chriseth) [reply libyul/optimiser/ForLoopConditionIntoBody.cpp]: Yes, but I'm not sure if that would be optimized at this stage.

Comment (ekpyron) [reply libyul/optimiser/ForLoopConditionIntoBody.cpp]: If we keep it like this, we should document both cases in the header.

File: libyul/optimiser/ForLoopConditionIntoBody.cpp (hunk 2 above)
Comment (chriseth) [reply libyul/optimiser/ForLoopConditionIntoBody.cpp]: You mean we could also ignore what happens after? right.

File: libyul/optimiser/ForLoopConditionOutOfBody.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ *   if iszero(c) { break }
+ *   ...
+ * }
+ *
+ * will be turned into
+ *
+ * for { ... } c { ... } {
+ *   ...
+ * }
+ *
```
Comment (ekpyron) [inline libyul/optimiser/ForLoopConditionOutOfBody.h:40]: and ``for { ... } 1 { ... } { if c  { break } ... }`` will be turned into ``for { ... } iszero(c) { ... } { ... }``

**Review by ekpyron — COMMENTED** [review]
I think the incomplete header docs is the only thing left here.

File: test/tools/yulopti.cpp [hunk 7]
```diff
@@ -43,6 +43,7 @@
 #include <libyul/optimiser/ExpressionInliner.h>
 #include <libyul/optimiser/FullInliner.h>
 #include <libyul/optimiser/ForLoopConditionIntoBody.h>
+#include <libyul/optimiser/ForLoopConditionOutOfBody.h>
```
Comment (leonardoalt) [inline test/tools/yulopti.cpp:46]: Not used?

File: test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul [hunk 8]
```diff
@@ -465,21 +465,19 @@
 //         let _1 := 0
 //         let _2 := mload(_1)
 //         let pos := 0x20
-//         let pos_1 := pos
 //         let length := mload(_2)
 //         mstore(pos, length)
 //         pos := 64
-//         let srcPtr := add(_2, pos_1)
+//         let srcPtr := add(_2, 0x20)
```
Comment (leonardoalt) [inline test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul:471]: Why?

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul:471]: This is an effect of the literal rematerializer. I couldn't tell which version is "better optimized".

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**